File: harness/dump.js
// Per-organism state dump — the port's bisection tool.
//
// When the raw fingerprint diverges, this says WHICH organism went wrong first, and its species
// says which branch of the tick to read. Slot order is the processing order, so the first
// differing line is the first divergence.
//
//   node harness/dump.js [ticks] [seed] [mutation 0|1]
const path = require("path");
const CORE = process.env.MC_CORE || path.join(__dirname, "..", "dist", "core.js");
const C = require(CORE);
const { W, P } = C;

const TICKS = parseInt(process.argv[2] || "1", 10);
const SEED = parseInt(process.argv[3] || "11", 10);
const MUT = process.argv[4] === "1";

const buf = Buffer.alloc(8);
const h = d => { buf.writeDoubleBE(d); return buf.toString("hex"); };

P.mutation = MUT;
C.resetWorld(); C.initWorld(SEED);
for (let t = 0; t < TICKS; t++) C.step();

for (let i = 0; i < W.n; i++){
  if (!W.alive[i]) continue;
  console.log([i, W.sp[i], h(W.x[i]), h(W.y[i]), h(W.en[i]), h(W.mn[i]), h(W.pr[i]),
    h(W.vx[i]), h(W.vy[i]), h(W.hd[i]), W.cy[i], W.gr[i], W.handle[i], W.cd[i], W.bst[i], W.pc[i], W.flee[i]].join(" "));
}

File: rust/microcosm-core/src/bin/dump.rs
//! Per-organism state dump — the Rust side of `harness/dump.js`. Same columns, same order.
//!
//!   cargo run --release --bin dump -- [ticks] [seed] [mutation 0|1]

use microcosm_core::Sim;

fn h(v: f64) -> String {
    format!("{:016x}", v.to_bits())
}

fn main() {
    let a: Vec<String> = std::env::args().skip(1).collect();
    let ticks: usize = a.first().and_then(|s| s.parse().ok()).unwrap_or(1);
    let seed: i32 = a.get(1).and_then(|s| s.parse().ok()).unwrap_or(11);
    let mutation = a.get(2).map(|s| s == "1").unwrap_or(false);

    let mut sim = Sim::new();
    sim.p.mutation = mutation;
    sim.reset_world();
    sim.init_world(Some(seed), None);
    for _ in 0..ticks {
        sim.step();
    }
    let w = &sim.w;
    let mut out = String::new();
    for i in 0..w.n_slots() {
        if w.alive[i] == 0 {
            continue;
        }
        out.push_str(&format!(
            "{} {} {} {} {} {} {} {} {} {} {} {} {} {} {} {} {}\n",
            i,
            w.sp[i],
            h(w.x[i] as f64),
            h(w.y[i] as f64),
            h(w.en[i] as f64),
            h(w.mn[i] as f64),
            h(w.pr[i] as f64),
            h(w.vx[i] as f64),
            h(w.vy[i] as f64),
            h(w.hd[i] as f64),
            w.cy[i],
            w.gr[i],
            w.handle[i],
            w.cd[i],
            w.bst[i],
            w.pc[i],
            w.flee[i]
        ));
    }
    print!("{}", out);
}
